cvmfs/export_plugin/fs_traversal.cc

```diff
@@ -0,0 +1,793 @@
+
+  // Touch is atomic, if it fails something else will write file
+  if (!dest->touch(dest->context_, src_st)
+    || (dest_st->cvm_checksum != NULL
+      && should_write_anyway(dest, src_st, dest_st)) ) {
+    char *src_ident = src->get_identifier(src->context_, src_st);
+    if (num_parallel_) {
+      FileCopy next_copy(src_ident, strdup(dest_data));
+
+      WritePipe(pipe_chunks[1], &next_copy, sizeof(next_copy));
+      atomic_inc64(&copy_queue);
+    } else {
+      if (!copyFile(src, src_ident, dest, dest_data, pstats)) {
+        LogCvmfs(kLogCvmfs, kLogStderr,
+          "Failed to copy %s->%s : %d : %s",
+          entry, dest_data, errno, strerror(errno));
+        errno = 0;
+        result = false;
+      }
+      free(src_ident);
+      pstats->Lookup(SHRINKWRAP_STAT_FILE_COUNT)->Inc();
+    }
+  } else {
+    pstats->Lookup(SHRINKWRAP_STAT_DEDUPED_FILES)->Inc();
+    pstats->Lookup(SHRINKWRAP_STAT_DEDUPED_BYTES)->Xadd(src_st->st_size);
+  }
+
+  // Should probably happen in the copy function as it is parallel
+  // Also this needs to be separate from copyFile, the target file
+  // could already exist and the link needs to be created anyway.
+  if (result && dest->do_link(dest->context_, entry, dest_data)) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+      "Failed to link %s->%s : %d : %s",
+      entry, dest_data, errno, strerror(errno));
+    errno = 0;
+    result = false;
+  }
+  free(dest_data);
+  return result;
+}
+
+bool handle_dir(
+  struct fs_traversal *src,
+  struct cvmfs_attr *src_st,
+  struct fs_traversal *dest,
+  struct cvmfs_attr *dest_st,
+  const char *entry
+) {
+  if (dest->do_mkdir(dest->context_, entry, src_st)) {
+    if (errno == EEXIST) {
+      errno = 0;
+      if (dest->set_meta(dest->context_, entry, src_st)) {
+        LogCvmfs(kLogCvmfs, kLogStderr,
+          "Traversal failed to set_meta %s", entry);
+        return false;
+      }
+    } else {
+      LogCvmfs(kLogCvmfs, kLogStderr,
+        "Traversal failed to mkdir %s : %d : %s",
+        entry, errno, strerror(errno));
+      return false;
+    }
+  }
+  return true;
+}
+
+void add_dir_for_sync(const char *dir, bool recursive) {
+  dirs_.push_back(new RecDir(strdup(dir), recursive));
+}
+
+bool Sync(
+  const char *dir,
+  struct fs_traversal *src,
+  struct fs_traversal *dest,
+  bool recursive,
+  perf::Statistics *pstats,
+  bool do_fsck) {
+  int cmp = 0;
+
+  char **src_dir = NULL;
+  size_t src_len = 0;
+  ssize_t src_iter = -1;
+  char * src_entry = NULL;
+  struct cvmfs_attr *src_st = cvmfs_attr_init();
+  list_src_dir(src, dir, &src_dir, &src_len);
+
+  char **dest_dir = NULL;
+  size_t dest_len = 0;
+  ssize_t dest_iter = -1;
+  char * dest_entry = NULL;
+  struct cvmfs_attr *dest_st = cvmfs_attr_init();
+  dest->list_dir(dest->context_, dir, &dest_dir, &dest_len);
+  qsort(dest_dir, dest_len, sizeof(char *),  strcmp_list);
+
+  // While both have defined members to compare.
```

Comment:
In general make sure the comment align with the current version of the loop. Some of these are from the original implementation.